llvm/lib/Transforms/Yk/ModuleClone.cpp

```diff
@@ -70,13 +70,54 @@ struct YkModuleClone : public ModulePass {
     }
   }
 
+  /**
+   * This function iterates over all functions in the `FinalModule`.
+   * If cloned function calls are identified within the original function
+   * instructions, they are redirected to the original function instead.
+   *
+   * **Example Scenario:**
+   * - Function `f` calls function `g`.
+   * - Function `g` is cloned as `__yk_clone_g`.
+   * - Function `f` is not cloned because its address is taken.
+   * - As a result, function `f` calls `__yk_clone_g` instead of `g`.
```

Comment:
IIUC the idea here is "we need to be able to trace `f`, but its address has taken, so we will always call `f` with tracing calls inserted"?

Reply:
More like:
We need to be able to trace `f`, but its address has taken.
We need to make sure that any functions called from `f` are also non-cloned functions (functions that we can trace).

Reply:
Updated tests :point_right: https://github.com/ykjit/ykllvm/pull/216/commits/0f3e89c080eaefa72b2832a3f34851b325c908b0

Reply:
> As a result, function `f` calls `__yk_clone_g` instead of `g`.

Is that what's already happening or what this function does? I feel like there's another bullet point missing.

Also, please remind me which functions have the `yk_record_block` calls, the original or the cloned ones?

Reply:
> Is that what's already happening or what this function does? I feel like there's another bullet point missing.

The comment example scenario is about what's currently happening (before this function in this PR)
I can update it to make it clearer though.

Reply:
>Also, please remind me which functions have the yk_record_block calls, the original or the cloned ones?
`yk_record_block` are part of the original function blocks. You can also see it in the `llvm/test/Transforms/Yk/ModuleClone.ll` test.

Reply:
I think what's happening here, is that we rename all functions in the cloned module before linking.
But then some of them are not renamed cause their address is taken, while sill referencing functions that were renamed.

let's say we have:
```
func a { }

// function which address is taken
func b { 
  a();
}
```
After calling `renameFunctions` on a cloned module we have:
```
func __yk_clone_a { }

// let b a function which address is taken
func b { 
  __yk_clone_a();
}
```
function `a` was renamed but function `b` wasn't (because address is taken).

And then when we link both modules,`b` from cloned module overrides the original `b` definition.

Reply:
Or maybe question here is how these call sites in the function that is not cloned got renamed in the first place?

Reply:
Looks like `F.setName` is actually updating the callsites as well as the function name.
I can see:
Before `F.setName` is called for `luaL_checkany` function:
```
21:                                               ; preds = %13
  tail call void @luaL_checkany(ptr noundef %0, i32 noundef 1) #77
  ...
  unreachable
```
And just after `F.setName` is called for `luaL_checkany` function:
```
21:                                               ; preds = %13
  tail call void @__yk_clone_luaL_checkany(ptr noundef %0, i32 noundef 1) #77
  ....
  unreachable
```

Reply:
> Is that what's already happening or what this function does? I feel like there's another bullet point missing.

Updated comment for clarity :point_right: 34629475e81902b127aa8da1b3fe45e431526d1f